girder/models/user.py

```diff
@@ -156,9 +179,33 @@ def authenticate(self, login, password):
         if user is None:
             raise AccessException('Login failed.')
 
+        # Validate OTP preconditions
+        if self.hasOtp(user):
+            if otpToken is None:
+                raise AccessException(
+                    'User authentication must include a one-time password '
```

Comment:
@girder/developers There are security implications to what the `AccessException` error message is. Previously, regardless of whether a user was not found or a password validation failed, the message `'Login failed.'` was returned. After this change, if a user is found who has OTP enabled, but no OTP is sent, the longer message `'User authentication must include a one-time password (typically in the "Girder-OTP" header).'` is send with the 401 error. This 'leaks' two pieces of information:
1. A user of that login / email addresses exists
2. That user has OTP enabled
If we want to consider any of this information to be secret, we can make some trade-offs in the order of the the various checks to leak some different information about password correctness instead. Unfortunately, to support the feature where not every user is prompted to enter an OTP token, we need to leak at least some information about the user existing and having OTP enabled.

Feel free bring up this with me directly, and I'll explain in more detail.